### ISSUE TITLE
[ABY] std::bad_alloc error on running dot example

### ISSUE THREAD

**Issue by bhatuzdaname** [mpc-msri/EzPC#107]
https://github.com/mpc-msri/EzPC/blob/master/EzPC/README.md#developers-manually-compiling-ezpc

On running the dot_product example we see the following error:
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
[1]    14753 abort      ./dot_product0 -r 1

**Comment by jaskiratsingh2000** [comment]
Hi @bhatuzdaname I am super interested to work on this. Just need bit of clarity before moving ahead directly. This issue is being arises when we manually compile `"./ezpc.sh dot_product.ezpc --bitlen 32"` it then creates the dot_product.cpp file and that is what you are trying to refer here. Right?

This error is generally referring to memory shortage or variables falling out of the main() function and if this is the case then it generally means that the changes/error has to be resolved within the **`https://github.com/mpc-msri/EzPC/blob/master/EzPC/EzPC/dot_product.ezpc`** file. Correct me if I am wrong?

Looking forward to hearing from you! Thanks :)

**Comment by bhatuzdaname** [comment]
Correct, when you compile the generated dot_product0.cpp with compile_aby.sh, the generated program on running gives the above error. You are right on track about it being a memory issue. You'll need to figure out where exactly we run into the issue.

**Comment by jaskiratsingh2000** [comment]
Thank you very much @bhatuzdaname for your response. I really appreciate that. 🙌
Well, just an another question and I think then I am good to go - So when we generate a file with this following command - `./ezpc.sh dot_product.ezpc --bitlen 32` it generates the file dot_product.cpp, so my question is that is there some folder or any directory where this generated file exist?

I am just new to it so never mi9nd. For any further queries I'll write up in this issue itself. Hope that would be fine! Thanks 😊

**Comment by bhatuzdaname** [comment]
Right that's the dot_product.ezpc you will have to compile. You will have to generate the cpp file yourself by running the ezpc.sh command. Follow https://github.com/mpc-msri/EzPC#setup to setup the repository.

**Comment by jaskiratsingh2000** [comment]
Thanks! Will get back to you soon.

**Comment by jaskiratsingh2000** [comment]
Hi @bhatuzdaname I successfully setup the repository on my local machine but while I was running the "`bash compile_aby.sh dot_product0.cpp`" it gave me the compilation error and this error actually seems to be different what you mentioned.

**Command ran:**
```
bash compile_aby.sh dot_product0.cpp
```

**Error received:**
```
./ABY_example/millionaire_prob_test.cpp:23:10: fatal error: ../../abycore/aby/abyparty.h: No such file or directory
   23 | #include "../../abycore/aby/abyparty.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from ./ABY_example/common/ezpc.h:27,
                 from dot_product0.cpp:50:
./ABY_example/common/millionaire_prob.h:22:10: fatal error: ../../../abycore/circuit/booleancircuits.h: No such file or directory
   22 | #include "../../../abycore/circuit/booleancircuits.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Compilation failed

```

I am not sure what's causing it but seems like there is the path issue and I can't even find those named files in my structure. Let me know if you have any idea about this. Thanks!

**Comment by bhatuzdaname** [comment]
I pushed a commit, could you pull and try again?

**Comment by jaskiratsingh2000** [comment]
Thanks @bhatuzdaname It worked!

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname Another question - if I make the changes to the **dot_product.ezpc** file then I have to compile it using the `**./ezpc.sh dot_product.ezpc -bitlen 32**` which generates the dot_product.cpp file then what does this means ./compile_aby.sh dot_product0.cpp? If I have to make changes and test what exactly do I have to run?

**Comment by bhatuzdaname** [comment]
ezpc.sh converts code written in the EzPC language to c++ code which relies on the ABY library. You can open and see what the compile_aby.sh script does. It basically compiles the generated c++ code and links it with ABY, setting the appropriate paths in the process. you can directly make changes to the cpp file and compile with compile_aby.sh. To run the generated program, see the instructions here https://github.com/mpc-msri/EzPC/blob/master/EzPC/README.md#running-ezpc-with-aby

**Comment by jaskiratsingh2000** [comment]
Thanks @bhatuzdaname It is working. When I try to print something within the generate .cpp file it doesn't accept like I am trying to print the std::cout << "comment" but it is not accepting this what could be an issue? Since this is the only way we can try debugging for this issue cause.

also, I checked that issue is only within the generated the .cpp file rest .ezpc file is totally okay. Do you have any hint that where can the cause be so that it would become easy to locate and work?

**Comment by bhatuzdaname** [comment]
could you please send the error you get after adding the cout statement? I
don't understand what you mean by not accepting..

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname Oh I mean to say that when I am trying to write "cout" within the **dot_product0.cpp** file it is not accepting that thing and hence throwing the following error:

```
dot_product0.cpp:49:6: error: ‘cout’ in namespace ‘std’ does not name a type
   49 | std::cout << "commented" << endl;
      |      ^~~~
In file included from dot_product0.cpp:7:
/usr/include/c++/9/iostream:61:18: note: ‘std::cout’ declared here
   61 |   extern ostream cout;  /// Linked to standard output
      |                  ^~~~
Compilation failed
```

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname I resolved the above error. The thing is that the `**std::cout**` has to be within the function and since there was no function in the generated **dot_product.cpp** file hence It works only when we put that inside any other function. So nevermind!

@bhatuzdaname My quick question is that what does ABYParty is linked in? Where can I file that is linked as ABYParty within the generated dot_product.cpp file? Can you please let me know?

Also I made some necessary changes within the .cpp file but I see that it is not part of this file structure officially howe can I resolve that? Looking forward to hearing from you! Thanks

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname Actually I have resolved this whole issue. Let me try explaining you!

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname I just have resolved this whole issue. I just want to know that I was able to do that by making changes to the generated .cpp file so how to commit for that since that is not part of an official file directory?

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname Are you around? Please let me know once.

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname I am still waiting on to hear back from you. As until I can get a better clarity oin this I can't commit my changes. So please let me know @bhatuzdaname Would highly appreciate that. Thanks

> @bhatuzdaname I just have resolved this whole issue. I just want to know that I was able to do that by making changes to the generated .cpp file so how to commit for that since that is not part of an official file directory?

**Comment by bhatuzdaname** [comment]
Hi @jaskiratsingh2000 , could you elaborate on how you fixed the issue in the generated .cpp file? then I can see the best way of applying that fix.

**Comment by jaskiratsingh2000** [comment]
Hi @bhatuzdaname Sure! 
Few minor fixes were required to resolve as I made the Pull Request for that above. Below are the steps to reproduce the fix.

* Compiling the EzPC file which will generate the **dot_product.cpp** file
```
./ezpc.sh dot_product.ezpc --bitlen 32
```

* So within the dot_product.cpp file. refer to the line no 377 which looks like the following 
```
party = new ABYParty(role_param, address, port, seclvl, bitlen, nthreads, mt_alg, 520000000);
```
Here to note that since this is doing a dynamic allocation so replace the above line with the following line:
```
ABYParty* party = new ABYParty(role_param, address, port, seclvl, bitlen, nthreads, mt_alg, 65536);
```

The reason to replace 520000000 with the 65536 is because within the ABY Project it has reservegates which has the deafult value of 65536.
https://github.com/encryptogroup/ABY/blob/d8e69414d091cafc007e65a03ef30768ebaf723d/src/abycore/aby/abyparty.h#L43-L44

* Compiling the EzPC with the ABY
```
./compile_aby.sh dot_product0.cpp 
``

* Running the compiled file
```
./dot_product0 -r 0
```

@bhatuzdaname I would be super happy and would like to create a PR for this myself. So can you please tell me that where exactly to commit for this? 
Looking forward to hearing from you!

Thanks

**Comment by bhatuzdaname** [comment]
Great job @jaskiratsingh2000 , thats the fix. So since the ezpc compiler generate this cpp file, we will change the ezpc compiler itself. 
In line 742 of codegen.ml, could you change 520000000 to 4000000 ?
The to rebuild the ezpc compiler do
```eval `opam config env`
make```
and to test, recompile with ezpc.sh and compile_aby.sh and verify if it doesnt crash. If it all works please send a PR with the codegen.ml changes.

**Comment by jaskiratsingh2000** [comment]
@bhatuzdaname Yes, everything is working fine after making changes. I just have a question that what is happening after we are changing the 520000000 to 4000000 and what does it number actually mean here in your opinion? Also What does the codegen.ml contains?

**Comment by bhatuzdaname** [comment]
In multi party computation, we lower computations to arithmetic/boolean circuits. In ABY we need to reserve a no. of gates to perform the computation. I took 4000000 from the largest example in the ABY source code.

The codegen.ml file belongs to the ezpc compiler and is responsible for the [code generation](https://en.wikipedia.org/wiki/Code_generation_(compiler)) phase. It visits the AST and outputs code for a given ezpc program.

**Comment by jaskiratsingh2000** [comment]
Well @bhatuzdaname This is super interesting. Thank you so much for your time on this specific issue. I really appreciate that.

Are there any similar kind of beginner-friendly issues which I can hop on to as this project made me more interested and I am super interested to contribute further. 👍

**Comment by drunkenlegend** [comment]
Issue closed by PR #109
